core/embed/rust/src/ui/model_tt/component/keyboard/passphrase.rs

```diff
@@ -252,20 +288,53 @@ impl Component for Input {
         let text_baseline = self.area.bottom_left() - TEXT_OFFSET;
         let text = self.textbox.content();
 
+        // Find out how much text can fit into the textbox.
+        let text_to_display: String<MAX_LENGTH> =
+            if style.font.text_width(text) <= self.area.width() {
+                // It can fit the whole passphrase, no need to do anything.
+                String::from(text)
+            } else {
+                // Passphrase is longer, showing its right end with ellipsis at the beginning.
+                let ellipsis = "...";
+                let ellipsis_width = style.font.text_width(ellipsis);
+                let awailable_width_for_text = self.area.width() - ellipsis_width;
+
+                // Finding out how many characters will fit in, starting from the right end.
```

Comment:
Nit: turning this into a method on `Font` like
```rust
fn longest_suffix(self, width: i16, text: &str) -> &str
// or
fn longest_suffix(self, width: i16, text: &str) -> usize
```
would make this a bit more readable and also potentially reusable.

Also there doesn't seem to be a need to build a string out of this, you can just display the two pieces of text directly?

Reply:
Good idea with creating a `Font` method for this.

With displaying two pieces of text, how can I do it without needing to call the `display::text` two times? (As that would create a lot of extra logic - sometimes calling it once, sometimes twice.)

Reply:
`Font` method done in [6e21a58](https://github.com/trezor/trezor-firmware/pull/2544/commits/6e21a58734413143eb851a2489463dfdc62e8f74)

Reply:
Why do you want to avoid calling it twice? Was thinking of something along the lines of:

```rust
let text_to_display = if fits_whole {
    text
} else {
    let ellipsis_width = ...;
    display_text(text_baseline, "...");
    text_baseline += ellipsis_width; // needed for paint_pending_marker
    text_suffix
}

display::text(text_baseline, text_to_display);
```

Reply:
Tried as an experiment in [2ef1b2a](https://github.com/trezor/trezor-firmware/pull/2544/commits/2ef1b2a7735b778d1a65cd4da17bfeae1b55b64c) - not sure if I like it, as it meant quite a lot of changes. But it removes the need for `heapless::String`, which is certainly helpful